### PR TITLE
Add YearRange and genDurationOf, harden tests slightly

### PR DESCRIPTION
**Pull request by sloshy**
Tests have been a little flaky recently (and possibly for a while, who knows) so I added a few changes to help make maintaining this project easier:

* `YearRange` - a new feature, an implicit parameter for determining a "realistic" range of years to generate values between. Not a more fine-grained range, just a set-it-and-forget-it type setting for controlling generated values.
* `genDurationOf` - a controllable version of `genDuration` that allows you to specify the min and max millis difference.
* Finally some changes to the `GenJdk8Properties` test suite, including:
  * Setting a `YearRange` between 0 and 10000, to avoid extremities at the weird ends of time as much as possible.
  * Adding a scenario testing for where dates literally cannot be normalized any further due to an early-morning time-zone change.
  * Tweaking the `genDateTimeWithinRange` spec to:
    * Use `genDurationOf` for more realistic duration differences.
    * Normalize the lower bound before comparison due to normalization sometimes giving you a lower-than-expected value (e.g. normalizing for `year` will give you the first nanosecond of the year, which is sometimes lower than your lower bound)

### CODE REVIEW

**Review by diesalbla — DISMISSED** [review]

**Comment by sloshy** [comment]
Sorry for re-requesting review @diesalbla, I realized that the timezone change detection wasn't very good, still failed, and had to rethink it. Thanks regardless!

**Review by juanpedromoreno — DISMISSED** [review]

File: modules/scalacheck-toolbox-datetime/src/main/scala/com/fortysevendeg/scalacheck/datetime/YearRange.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+trait YearRange {
+  val min: Int
+  val max: Int
+}
+
+object YearRange {
+
+  /** The default range of years. Set to a safe subset shared by both `java.time` and `org.joda.time`. */
+  implicit def default: YearRange = new YearRange {
```
Comment (cb372) [inline modules/scalacheck-toolbox-datetime/src/main/scala/com/fortysevendeg/scalacheck/datetime/YearRange.scala:11]: could be a `val`

File: modules/scalacheck-toolbox-datetime/src/main/scala/com/fortysevendeg/scalacheck/datetime/YearRange.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    val min: Int = -292275054
+    val max: Int = 292278993
+  }
+
+  /** Defines a year range between 1970 and the max year you specify. */
+  def epochTo(year: Int) = new YearRange {
+    override val min = 1970
+    val max          = year
```
Comment (cb372) [inline modules/scalacheck-toolbox-datetime/src/main/scala/com/fortysevendeg/scalacheck/datetime/YearRange.scala:19]: nit: why is one marked with `override` but not the other? Same goes for the method below.

File: modules/scalacheck-toolbox-datetime/src/main/scalajvm/com/fortysevendeg/scalacheck/datetime/joda/GenJoda.scala [hunk 3]
```diff
@@ -23,15 +23,16 @@ import scala.util.Try
 import org.scalacheck.Gen
 import org.scalacheck.Arbitrary
 import org.joda.time._
+import com.fortysevendeg.scalacheck.datetime.YearRange
 
 /**
  * Generators specific for Joda time.
  */
 trait GenJoda {
```
Comment (cb372) [inline modules/scalacheck-toolbox-datetime/src/main/scalajvm/com/fortysevendeg/scalacheck/datetime/joda/GenJoda.scala:31]: Random question, unrelated to your PR. Can't we remove support for joda-time? Does anyone still use that?

**Review by cb372 — APPROVED** [review]
This all seems sensible 👍

**Review by noelmarkham — APPROVED** [review]

File: modules/scalacheck-toolbox-datetime/src/main/scala/com/fortysevendeg/scalacheck/datetime/YearRange.scala (hunk 1 above)
Comment (sloshy) [reply modules/scalacheck-toolbox-datetime/src/main/scala/com/fortysevendeg/scalacheck/datetime/YearRange.scala]: Fixed.

File: modules/scalacheck-toolbox-datetime/src/main/scala/com/fortysevendeg/scalacheck/datetime/YearRange.scala (hunk 2 above)
Comment (sloshy) [reply modules/scalacheck-toolbox-datetime/src/main/scala/com/fortysevendeg/scalacheck/datetime/YearRange.scala]: An earlier version had min pre-set as `-max`, and I just forgot to remove the override keyword. My bad. Fixed.

**Comment by sloshy** [comment]
@cb372 addressed your comments, thanks for the feedback!

**Pull request merged**